syncer/ddl.go

```diff
@@ -29,147 +28,95 @@ import (
-					isDDL:    false,
-				}, nil
-			}
-		}
-		return parseDDLResult{
-			stmt:     nil,
-			needSkip: false,
-			isDDL:    false,
-		}, terror.Annotatef(terror.ErrSyncUnitDMLStatementFound.Generate(), "query %s", sql)
-	default:
-		// BEGIN statement is included here.
-		// let sqls be empty
-		return parseDDLResult{
-			stmt:     nil,
-			needSkip: false,
-			isDDL:    false,
-		}, nil
+		return nil, nil
 	}
+	return stmts[0], nil
 }
 
-// splitAndFilterDDL will split multi-schema change DDL into multiple one schema change DDL due to TiDB's limitation.
-// the returned DDLs is split from `stmt`, or DDLs that online DDL component has saved before (which means the input
-// `stmt` is a RENAME statement which often generated by online DDL tools).
-// return @spilt sqls, @online ddl table names, @error.
-func (s *Syncer) splitAndFilterDDL(
-	ec eventContext,
-	p *parser.Parser,
-	stmt ast.StmtNode,
-	schema string,
-) (sqls []string, tableMap map[string]*filter.Table, err error) {
-	sqls, err = parserpkg.SplitDDL(stmt, schema)
+// processSplitedDDL processes SplitedDDLddl as follow step:
+// 1. track ddl whatever skip it except optimist shard ddl; (TODO: will implement in https://github.com/pingcap/dm/pull/1975)
+// 2. skip sql by filterQueryEvent;
+// 3. apply online ddl if onlineDDL is not nil:
+//    * specially, if skip, apply empty string;
+// 4. handle online ddl SQL by handleOnlineDDL.
+func (s *Syncer) processSplitedDDL(qec *queryEventContext, sql string) ([]string, error) {
+	ddlInfo, err := s.routeDDL(qec.p, qec.ddlSchema, sql)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
+	// TODO: add track ddl
 
-	statements := make([]string, 0, len(sqls))
-	tableMap = make(map[string]*filter.Table)
-	for _, sql := range sqls {
-		stmt2, err2 := p.ParseOneStmt(sql, "", "")
-		if err2 != nil {
-			return nil, nil, terror.Annotatef(terror.ErrSyncerUnitParseStmt.New(err2.Error()), "ddl %s", sql)
+	// get real tables before apply block-allow list
+	realTables := make([]*filter.Table, 0, len(ddlInfo.sourceTables))
+	for _, table := range ddlInfo.sourceTables {
+		realName := table.Name
+		if s.onlineDDL != nil {
+			realName = s.onlineDDL.RealName(table.Name)
 		}
+		realTables = append(realTables, &filter.Table{
+			Schema: table.Schema,
+			Name:   realName,
+		})
+	}
 
-		tables, err2 := parserpkg.FetchDDLTables(schema, stmt2, s.SourceTableNamesFlavor)
-		if err2 != nil {
-			return nil, nil, err2
-		}
+	shouldSkip, err := s.skipQueryEvent(realTables, ddlInfo.stmt, sql)
+	if err != nil {
+		return nil, err
+	}
+	if shouldSkip {
+		metrics.SkipBinlogDurationHistogram.WithLabelValues("query", s.cfg.Name, s.cfg.SourceID).Observe(time.Since(qec.startTime).Seconds())
+		qec.tctx.L().Warn("skip event", zap.String("event", "query"), zap.String("statement", sql), zap.String("schema", qec.ddlSchema))
+		// when skip ddl, apply empty ddl for onlineDDL
+		// otherwise it cause an error once meet the rename ddl of onlineDDL
+		sql = ""
```

Comment:
What do we the onlineDLL module want to do with an empty SQL. Can we directly return nil here 🤔